### PR TITLE
ENH: Support for Riesz-transformed LoG, Riesz-transformed Simoncelli, and Simoncelli wavelet filter

### PR DESCRIPTION
**Pull request by maksymfritsak**
Introduction of the last missing filter from IBSI II that still has compliance reports: Riesz-transformed LoG, Riesz-transformed Simoncelli, and Simoncelli

### CODE REVIEW

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by hubertgabrys** [comment]
Great improvement. Many thanks @maksymfritsak . Could you please specify in the PR description which new filter types are now supported?

**Comment by maksymfritsak** [comment]
Thanks, @hubertgabrys!

Sure, I still want to fix a few borderline issues. It should be ready by this evening, so we can either include it in the August update or move it to September—up to you.

File: zrad/gui/filt_tab.py [hunk 1]
```diff
@@ -764,6 +805,25 @@ def _filter_combo_box_changed(self, text):
...
+            self.riesz_order_field.show()
+            if text == 'Riesz-transformed LoG':
+                self.log_filter_sigma_label.show()
+                self.log_filter_sigma_text_field.show()
+                self.log_filter_cutoff_label.show()
+                self.log_filter_cutoff_text_field.show()
+                self.structure_tensor_sigma_label.show()
+                self.structure_tensor_sigma_field.show()
+            else:
+                self.wavelet_filter_decomposition_level_combo_box.show()
```
[the remark this answers is not in the document]
Comment (maksymfritsak) [reply zrad/gui/filt_tab.py:833]: Implemented already

File: zrad/gui/filt_tab.py [hunk 2]
```diff
@@ -764,6 +805,25 @@ def _filter_combo_box_changed(self, text):
             self.gabor_rotinv_box.show()
             self.gabor_ortho_box.show()
 
+        elif text in ('Riesz-transformed LoG', 'Simoncelli'):
+            if self.wavelet_filter_response_map_connected:
+                self.filter_dimension_combo_box.currentTextChanged.disconnect()
+                self.wavelet_filter_response_map_connected = False
+            self._hide_existing_filter_parameter_widgets()
+            self.padding_type_combo_box.show()
```
[the remark this answers is not in the document]
Comment (maksymfritsak) [reply zrad/gui/filt_tab.py:817]: No longer relevant

**Comment by hubertgabrys** [comment]
No stress, let's make a September release with this PR and #282.

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

File: zrad/filtering/spatial.py [hunk 3]
```diff
@@ -141,6 +143,143 @@ def _apply_array(self, img):
...
+    def _boundary_aware_riesz_transform(self, image, order):
+        """Apply the Riesz transform without imposing unintended periodicity."""
+        if self.padding_type == 'wrap':
+            return self._riesz_transform(image, order)
+
+        if self.padding_type in ('constant', 'nearest'):
+            # These boundary modes do not have a cosine-transform
+            # representation. Extend the domain according to the selected
+            # mode before applying the periodic Riesz transform.
+            padding = tuple((size // 2, size - size // 2) for size in image.shape)
```
[the remark this answers is not in the document]
Comment (maksymfritsak) [reply zrad/filtering/spatial.py:247]: This is a price I am ready to pay for a fully IBSI compliant implementation. @hubertgabrys what do you think?

**Comment by maksymfritsak** [comment]
@hubertgabrys 

Potentially completed

**Comment by maksymfritsak** [comment]
@codex review

File: zrad/filtering/spatial.py [hunk 4]
```diff
@@ -141,6 +143,166 @@ def _apply_array(self, img):
...
+                response += (
+                    target_coefficient
+                    / np.sqrt(2.0)
+                    * (
+                        first_direction[..., row] * second_direction[..., column]
+                        + first_direction[..., column] * second_direction[..., row]
+                    )
+                    * self._boundary_aware_riesz_transform(log_response, order)
+                )
+        return response
```
[the remark this answers is not in the document]
Comment (maksymfritsak) [reply zrad/filtering/spatial.py:372]: The safest IBSI-aligned solution is to reject mixed aligned orders, since IBSI’s tested aligned configuration uses (0,2,0).

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@hubertgabrys, looks like we finally did it!

**Review by hubertgabrys — CHANGES_REQUESTED** [review]
The 3-D nearest/constant Fourier paths materialize an 8x padded volume, three dense coordinate grids, a dense radius/numerator/response, and full complex FFT buffers. On a representative 128x128x64 volume, nearest Simoncelli peaks at 1.44 GB and constant Riesz-LoG at 1.26 GB; ordinary clinical volumes can therefore exhaust a typical workstation.

Please retain the boundary extension but remove the avoidable full-grid allocations. A concrete implementation that preserves the current output is:

- Generate FFT coordinates with np.meshgrid(..., sparse=True).
- Use scipy.fft.rfftn/irfftn so real inputs only require a half-spectrum.
- Allocate one dense half-spectrum radial buffer and reuse it in place for the Simoncelli response or Riesz denominator.
- Multiply the filter and Riesz response directly into the FFT spectrum instead of allocating separate numerator, multiplier, and filtered-spectrum arrays.
- Preserve equivalence for odd Riesz orders by explicitly cancelling Nyquist bins where an odd number of Nyquist-axis powers contribute; this matches taking the real part of the existing full inverse FFT.

I prototyped this formulation and measured peaks of 536 MB for nearest Simoncelli and 493 MB for constant Riesz-LoG, reductions of 63% and 61%. Outputs matched the existing implementation to about 1e-15, IBSI configurations 8 and 9 passed, and configuration 10.b.1 had zero voxels outside the official tolerance.

**Review by hubertgabrys — CHANGES_REQUESTED** [review]
Request changes: extend once before composing LoG and Riesz

`RieszLoG._apply_array` first calls `LoG._apply_array(img)`, which evaluates the LoG under the selected boundary condition and returns only the original field of view. It then passes that cropped response to `_boundary_aware_riesz_transform`, which extends the domain a second time. For constant padding, this is not equivalent to applying the composed LoG–Riesz filter on a single extension of the input: the first step discards the LoG tail outside the field of view, even though that tail is required by the subsequent nonlocal Riesz transform. Although the discrepancy is caused at the boundary, the Fourier-domain Riesz operation can propagate it across the output.

The IBSI checks do not disprove this issue. Configuration 9.a gives identical results because its centered impulse has enough zero margin that the finite LoG support never reaches the boundary. In 9.b.1, the current and extend-once implementations differ by up to 0.145, but the IBSI acceptance tolerance is 0.846, so both pass. The extend-once result is also closer to the reference map (maximum error 0.267 versus 0.364 for the current implementation).

Please move boundary extension/cropping to the composition level. For constant and nearest modes, extend the original image once, evaluate LoG on that extended domain, apply the raw Riesz transform to the extended LoG response, and crop only the final result. Avoid calling a helper that independently reapplies the boundary condition between the two operators. Apply the same domain handling to each 2D slice and to the structure-tensor path. Wrap can operate directly on the periodic domain; reflect/DCT should likewise preserve one consistent extension model.

Please also add a focused regression test with nonzero structure within the LoG support radius of an image boundary. Compare the implementation against an explicitly extended-domain LoG-then-Riesz reference and assert agreement after the final crop. This covers the case that the current IBSI phantoms and tolerance do not distinguish.

**Comment by hubertgabrys** [comment]
Also, why was IBSI_II.zip modified?

**Comment by maksymfritsak** [comment]
> Also, why was IBSI_II.zip modified?

We did not have Pattern 1 phantom there before for the Phase I 10.b.1 test

**Comment by maksymfritsak** [comment]
> Also, why was IBSI_II.zip modified?

What’s weird is that it’s actually lighter than the previous .zip file. However, I checked the contents of both .zip files, and everything is there.

**Comment by hubertgabrys** [comment]
> > Also, why was IBSI_II.zip modified?
> 
> What’s weird is that it’s actually lighter than the previous .zip file. However, I checked the contents of both .zip files, and everything is there.

I've also checked it. Perhaps different compression level.

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@codex review

**Comment by maksymfritsak** [comment]
@hubertgabrys 

Thanks for your suggestions on this PR!

It’s finally ready for you to review :)

**Review by hubertgabrys — CHANGES_REQUESTED** [review]
Many thanks Maksym. Great work. I think we are good on logic of these changes. However, it seems the docs and docstrings are not entirely updated, no?

**Comment by maksymfritsak** [comment]
I've compared the 10A configuration across our results, MIRP, and the reported IBSI values. Interestingly, both our results and MIRP are in alignment with one of the reported IBSI results.

image

We are in agreement with Veneto Institute of Oncology: 

Mean                                           : -0.149
Variance                                      : 606.656
Skewness                                    : -0.098
Excess kurtosis                            : 9.451
Median                                         : 0.002
Minimum                                      : -260.199
10th percentile                             : -20.589
90th percentile                             : 20.359
Maximum                                     : 272.181
Interquartile range                       : 16.065
Range                                          : 532.379
Mean absolute deviation              : 14.732
Robust mean absolute deviation  : 7.085
Median absolute deviation           : 14.731
Coefficient of variation                 : -165.693
Quartile coefficient of dispersion : 332.930
Energy                                         : 75990038.613
Root mean square                       : 24.631


And MIRP results are similar to Qurit SERA: 

Mean                                         : -0.126
Variance                                    : 570.720
Skewness                                  : -0.123
Excess kurtosis                         : 13.442
Median                                      : 0.024
Minimum                                    : -294.976
10th percentile                           : -17.911
90th percentile                           : 17.970
Maximum                                   : 262.969
Interquartile range                      : 12.642
Range                                         : 557.945
Mean absolute deviation             : 13.261
Robust mean absolute deviation : 5.746
Median absolute deviation          : 13.260
Coefficient of variation                 : -189.041
Quartile coefficient of dispersion : 337.116
Energy                                         : 71488122.711
Root mean square                       : 23.890

**Comment by maksymfritsak** [comment]
@hubertgabrys 
Completed!

**Review by hubertgabrys — APPROVED** [review]
